r/R/array.R

```diff
@@ -187,7 +187,18 @@ Array$create <- function(x, type = NULL) {
     }
     return(out)
   }
-  vec_to_Array(x, type)
+  tryCatch(
+    vec_to_Array(x, type),
+    error = function(cnd) {
+      if (!is.null(type)) {
+        # try again and then cast
+        vec_to_Array(x, NULL)$cast(type)
+      } else {
+        signalCondition(cnd)
+      }
+    }
+  )
+
```

Comment:
Would this be considered a fix for [ARROW-11631](https://issues.apache.org/jira/browse/ARROW-11631)?
As @romainfrancois indicated https://github.com/apache/arrow/pull/11805#issuecomment-985352985 this would allow to create Arrays directly.

Reply:
It might not be as straightforward as initially thought. Catching the error and then casting has some unintended consequences. For example, some errors get muddled due to the fact they now pass through cast - while they are somewhat easy to solve with the C++ error message.

Before:
``` r
library(arrow, warn.conflicts = FALSE)
df <- tibble::tibble(x = 1:10, y = 1:10)
type <- struct(x = float64(), y = int16(), z = int32())
Array$create(df, type = type)
#> Error: Unknown: Number of fields in struct (3) incompatible with number of columns in the data frame (2)
```

<sup>Created on 2021-12-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>


After:
```r
library(arrow, warn.conflicts = FALSE)
df <- tibble::tibble(x = 1:10, y = 1:10)
type <- struct(x = float64(), y = int16(), z = int32())
Array$create(df, type = type)
#>Error: NotImplemented: Unsupported cast from struct<x: int32, y: int32> to struct using function cast_struct
/Users/dragos/Documents/arrow/cpp/src/arrow/compute/function.cc:215 DispatchBest(&inputs)
```

<sup>Created on 2021-12-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>


Same for a mismatch in names. The more informative C++ error message:
``` r
library(arrow, warn.conflicts = FALSE)
df <- tibble::tibble(x = 1:10, y = 1:10)
type <- struct(y = int16(), x = float64())
Array$create(df, type = type)
#> Error: Unknown: Field name in position 0 (y) does not match the name of the column of the data frame (x)
```

<sup>Created on 2021-12-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

becomes:
``` r
library(arrow, warn.conflicts = FALSE)
df <- tibble::tibble(x = 1:10, y = 1:10)
type <- struct(y = int16(), x = float64())
Array$create(df, type = type)
#> Error: NotImplemented: Unsupported cast from struct<x: int32, y: int32> to struct using function cast_struct
/Users/dragos/Documents/arrow/cpp/src/arrow/compute/function.cc:215 DispatchBest(&inputs)
```

<sup>Created on 2021-12-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
These ☝🏻 are both caught by failing unit tests.

Reply:
It looks like we need to somehow capture the error from `vec_to_Array()$cast(type)` and return the original condition. Looking into it.

Reply:
oh good point. Maybe casting automatically is not such a good idea then.

Reply:
Another approach would be to hint about `$cast()`: 

 - First try to do `vec_to_Array(x, type)` 
 - If this fails and `type` was given, there is a chance that `vec_to_Array(x, NULL)$cast(type)` would work. The code could attempt it. 
     - if we end up with an Array of `type` and then hint it, i.e. hint about doing `Array$create(x)$cast(type)
     - if not, just rethrow the original error

Hinting places the decision of using casting on the user

Reply:
_nods_ yeah I think auto-casting might end up being more confusing than helpful (though hinting that creating then casting manually might work is nice).

Did I miss in here what the blocker is to being able to create these directly / ARROW-11631?

Reply:
Hinting could be a stepping stone, but yeah creating directly would be preferable.

Reply:
The `tryCatch(...)` now looks like this and I think it's a step in the right direction:
```
tryCatch(
    vec_to_Array(x, type),
    error = function(cnd) {
      attempt <- try(vec_to_Array(x, NULL)$cast(type), silent = TRUE)
      if (!inherits(attempt, "try-error")) {
        message("Hint: You might want to try casting manually. Something like `Array$create(...)$cast(...)`")
      }
      stop(conditionMessage(cnd), call. = FALSE)
    }
  )
  ```

Reply:
Perhaps it's worth using `abort()` bullets: 

```r
abort(c(
  conditionMessage(cnd), 
  i = if (!inherits(attempt, "try-error")) "You might want to try casting manually. Something like `Array$create(...)$cast(...)`"
))
```

Reply:
@jonkeane not sure on the blocker / [ARROW-11631](https://issues.apache.org/jira/browse/ARROW-11631). For me it's lack of C++ skills 😅

Reply:
How could the error message be improved?

Reply:
Ah, sorry what I meant by "this message isn't great" is that having to have a (stop) message here isn't the best UI (but we're deferring ARROW-11631 since it's not trivial), so this is the next best thing.

The content of the message proposed here is good, it's clear + actionable and better than having people stuck like they are now.